### DISCUSSION TITLE
Multi-tenancy questions/issues, with bonus alembic (and SQLModel)

### DISCUSSION THREAD

**Discussion by tftio** [sqlalchemy/sqlalchemy#12139] [Usage Questions] [answered]
Hello, all!

Our application is a multi-tenant one -- there is a set of entities that live in a shared schema -- call it `shared` -- things that are application level (accounts, users, permissions, &c.). Then, for every customer, we create a new copy of a `tenant_template` schema that has customer-specific entities in it. I manage this in the application by parameterizing our database session: when you ask for a specific customer session, I set the `schema_translate_map` option on the engine:

```
pool = get_a_connection_pool_function_goes_here()

def get_tenant_engine(tenant_schema_name: str | None = None) -> Engine:
    """Specialize a DB engine to a single tenant."""

    if tenant_schema_name is not None:
        return pool.execution_options(
            schema_translate_map={
                None: tenant_schema_name,
            }
        )
    return pool
```

This works fine; I manage the above with FastAPI and it's dependency injection facility, so every method call we make has an implicit session built on that engine. OK, works fine. 

I also use `alembic`, with some customizations, and here's where the issue is. The problem I have is that if I want to use alembic's super helpful autogeneration, I have to manually add set the schema for all my models, *as well as any foreign key constraints*, run the revision generation, and then remove the schema from my models. 

I'd love to stop doing this; as the team grows, my fiddly approach here becomes a bottleneck. I guess one thing I could do is set the schema on the models to be the `TENANT_TEMPLATE_SCHEMA`, and then add that to the translation map? But, the foreign key constraints I have are for the most part hardcoded strings:

```
SHARED_SCHEMA: str = "shared_data"
TENANT_TEMPLATE_SCHEMA: str = "tenant_template"

class AccountInfo(SQLModel, table=True):
    """Miscellaneous account information."""

    __tablename__ = "account_info"  # pyright: ignore [reportAssignmentType]
    __table_args__ = (
        PrimaryKeyConstraint("account_name", name="account_info_pkey"),
        ForeignKeyConstraint(
            ["account_name"],
            [f"{SHARED_SCHEMA}.accounts.account_name"],
            name="account_info_account_name_fkey",
            ondelete="CASCADE",
            onupdate="CASCADE",
        ),
        ForeignKeyConstraint(
            ["default_dataset"],
            [f"{TENANT_TEMPLATE_SCHEMA}.datasets.dataset_name"], # this needs to point to the tenant-specific dataset table! 
            name="account_info_default_dataset_fkey",
            ondelete="SET NULL",
            onupdate="CASCADE",
        ),
       { "schema": TENANT_TEMPLATE_SCHEMA },
    )
    account_name: str = Field(primary_key=True, sa_type=CITEXT, nullable=False)
    default_dataset: str | None = Field(sa_type=CITEXT, nullable=True, default=None)
```

I don't mind doing this translation at runtime, for instance, walking the graph of models and modifying the schema and the references; but I don't know enough about the SQLAlchemy internals to know if there's a pre-existing hook or even a spot to write this functionality in. 

TIA for any light you can shed!

**Comment by zzzeek** [comment, marked as answer]
>  I have to manually add set the schema for all my models, as well as any foreign key constraints, run the revision generation, and then remove the schema from my models.

missing here is why do you have to do that?   this depends highly on what kind of database you're using but let's assume it's PostgreSQL, set the `search_path` for that autogenerate so that whatever schema you want to target is implicit.   See the recipe at https://alembic.sqlalchemy.org/en/latest/cookbook.html#rudimental-schema-level-multi-tenancy-for-postgresql-mysql-other-databases for a general approach to set up an alembic run to see a certain schema as default

**Reply by tftio** [reply]
It's postgres, yes. I have a separate affordance for running the migration against all the tenant schemas in the database; in the presence of a `-x everywhere` flag, the migration is written to iterate over all the existing tenants. But this looks more promising; I'll need to check it out in more depth. Thank you!
